File: tree/543_diameter_of_binary_tree.cc
#include <algorithm>

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

// 2024/01/07: 自己做出来了从上到下的版本, 没有做出来从下到上的版本.
// 2024/01/02: 在做了110的前提下可以自己写出来.
int res = 0;

int Depth(TreeNode* root) {
  if (root == nullptr) {
    return 0;
  }
  const int left_depth = Depth(root->left);
  const int right_depth = Depth(root->right);
  res = std::max(res, left_depth + right_depth);
  return std::max(left_depth, right_depth) + 1;
}

int DiameterOfBinaryTree(TreeNode* root) {
  Depth(root);
  return res;
}

File: tree/543_diameter_of_binary_tree.py
from typing import Optional
from tree.tree_node import TreeNode, create_tree_node, print_tree_node

diameter = 0


# 2026/05/18: 自己写出来了, 但是 height 调用重复导致超出时间限制, 应该先保存.
def height(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0
    left_height = height(root.left)
    right_height = height(root.right)
    diameter = max(diameter, left_height + right_height)
    return max(left_height, right_height) + 1


def diameter_of_binary_tree(root: Optional[TreeNode]) -> int:
    height(root)
    return diameter


root = [
    4,
    -7,
    -3,
    None,
    None,
    -9,
    -3,
    9,
    -7,
    -4,
    None,
    6,
    None,
    -6,
    -6,
    None,
    None,
    0,
    6,
    5,
    None,
    9,
    None,
    None,
    -1,
    -4,
    None,
    None,
    None,
    -2,
]
root = create_tree_node(root)
print_tree_node(root)
print(diameter_of_binary_tree(root))
